pkg/cli/build.go

```diff
@@ -0,0 +1,200 @@
+package cli
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/maps"
+
+	"github.com/wolfi-dev/wolfictl/pkg/dag"
+)
+
+func Build() *cobra.Command {
+	var archs []string
+	var dir, pipelineDir string
+	var jobs int
+	var dryrun bool
+	cmd := &cobra.Command{
+		Use:           "build",
+		SilenceErrors: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+
+			if jobs == 0 {
+				jobs = runtime.GOMAXPROCS(0)
+			}
+			jobch := make(chan struct{}, jobs)
+
+			newTask := func(pkg string) *task {
+				return &task{
+					pkg:         pkg,
+					dir:         dir,
+					pipelineDir: pipelineDir,
+					stdout:      cmd.OutOrStdout(),
+					stderr:      cmd.ErrOrStderr(),
+					archs:       archs,
+					dryrun:      dryrun,
+					done:        make(chan struct{}),
+					deps:        map[string]chan struct{}{},
+					jobch:       jobch,
+				}
+			}
+
+			if len(args) == 1 {
+				// Build only this one package.
+				t := newTask(args[0])
+				go t.start(ctx)
+				return t.wait(ctx)
+			}
+
+			pkgs, err := dag.NewPackages(os.DirFS(dir), dir, pipelineDir)
+			if err != nil {
+				return err
+			}
+			g, err := dag.NewGraph(pkgs)
+			if err != nil {
+				return err
+			}
+			if len(args) > 1 {
+				// If multiple args were passed, build only the subgraph based on these packages.
+				g, err = g.SubgraphWithRoots(args)
+				if err != nil {
+					return err
+				}
+			}
+			// Only return local packages
+			g, err = g.Filter(dag.FilterLocal())
+			if err != nil {
+				return err
+			}
+
+			// Only return main packages (configs)
+			g, err = g.Filter(dag.OnlyMainPackages(pkgs))
+			if err != nil {
+				return err
+			}
+
+			m, err := g.Graph.AdjacencyMap()
+			if err != nil {
+				return err
+			}
+
+			tasks := map[string]*task{}
+			for _, pkg := range g.Packages() {
+				if tasks[pkg] == nil {
+					tasks[pkg] = newTask(pkg)
+				}
+				for k, v := range m {
+					if strings.HasPrefix(k, pkg+":") {
```

Comment:
This could use a comment

Reply:
I'll squeeze this into #353 so I don't undo your approval.